### DISCUSSION TITLE
Choosing RWMutex

### DISCUSSION THREAD

**Discussion by zephyrtronium** [zephyrtronium/zephyrtronium.github.io#1] [Announcements]
https://zephyrtronium.github.io/articles/rwmutex.html

**Comment by weiwei-lin** [comment]
Nice article!

I wonder if the problem you described in [this section](https://zephyrtronium.github.io/articles/rwmutex.html#8.) also applies to regular `Mutex`. Given that `Mutex` also needs to do atomic update when there are a few threads just want to lock it for reads.

If so, the only disadvantage of `RWMutex` over `Mutex` in a read heavy load is just that `RWMutex` is larger and therefore is less cache friendly.

Am I right or did I miss something? Much appreciated!

**Reply by zephyrtronium** [reply]
I developed a response to this and then never actually posted it. Oops.

It's true that many threads using atomic CAS can serialize in a manner similar to what I described in that section. The difference is really in what happens when threads share caches. Let's use the same names and ordering as in that example, just swapping LOCK ADD for LOCK CMPXCHG, but let's add some more assumptions.

First, we'll assume that the mutex is actively contended and therefore its memory location is currently live in every cache. Second, we'll imagine that even and odd threads share L2 caches, and all processors share an L3 cache. (I think this would be a rather dated setup in modern processors, but it's still plausible.)

1. T0's CMPXCHG initiates first and updates the variable's value, which means that cache line is marked modified in T0's L1 cache and invalid in every other cache.
2. T1 gets to go next. Its read triggers T1's L1 cache to pull from the T1/3 L2 cache, which also doesn't have it. It will in turn pull from the L3 cache; this will repeat, triggering write-back from T0's L1, to the T0/2 L2 cache, and then to everyone's L3 cache.
3. T1's CMPXCHG executes, but since T0 is the lock owner, the CMPXCHG doesn't write to memory. Caches remain as they are.
4. T2 now goes, again triggering a pull from L2 cache, but this time it comes from the T0/2 L2 cache which already has the value due to T1's read. Only T2's L1 needs to update.
5. T2's CMPXCHG executes, again changing nothing.
6. T3's turn; it happens exactly like T2's.

In the RWMutex scenario, each thread forces the cache to synchronize at the L3 level, since LOCK ADD with a nonzero addend always updates the value. Even if we generously assume that writes to each level of cache have the same cost – I believe generally on the order of several dozen CPU cycles – the CMPXCHG approach saves on O(caches) such writes. T2 in particular spends half the time on Mutex than RWMutex given this setup, as will whichever thread eventually unlocks it. (And only one thread will unlock it, since only one thread locked it originally.)

So, it's correct to say that RWMutex is less cache friendly than Mutex, but there is more to it than just the object sizes. It has a direct time cost as well.

**Reply by weiwei-lin** [reply]
Thanks! That’s very helpful!
